### PR TITLE
Implement SCP Support

### LINKED ISSUE #151: Add support for SCP libssh API's
##### SUMMARY

Add support for SCP API's to be invoked from Ansible libssh connection plugin
```python
get_transport()
put()
get()
```

##### ISSUE TYPE

- Feature Request

### PR DESCRIPTION
**Pull request by Qalthos**
##### SUMMARY

Fixes https://github.com/ansible/pylibssh/issues/151
Adds scp support to pylibssh.

##### ISSUE TYPE

- Feature Pull Request

### CODE REVIEW

File: src/pylibsshext/scp.pyx [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+
+cdef class SCP:
+    def __cinit__(self, session):
+        self.session = session
+        self._libssh_session = get_libssh_session(session)
+
+    def __dealloc__(self):
+        if self._libssh_session is not NULL:
+            self._libssh_session = NULL
```
Comment (ganeshrn) [inline src/pylibsshext/scp.pyx]: Assigning NULL if memory is allocated doesn't seem right, might result in memory leaks. Is the `__dealloc__` required here? The deallocation of `self._libssh_session` should be taken care of by the session object.

File: src/pylibsshext/scp.pyx [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+            return libssh.SSH_OK
+
+    def get(self, remote_file, local_file):
+        cdef char *read_buffer
+
+        remote_file_b = remote_file
+        if isinstance(remote_file_b, unicode):
+            remote_file_b = remote_file.encode("utf-8")
+            remote_dir_b, filename_b = os.path.split(remote_file_b)
```
Comment (ganeshrn) [inline src/pylibsshext/scp.pyx]: Is this line required `remote_dir_b` and `filename_b` not used in function?

File: src/pylibsshext/scp.pyx [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+            size = libssh.ssh_scp_request_get_size(scp)
+            mode = libssh.ssh_scp_request_get_permissions(scp)
+
+            read_buffer = <char *>PyMem_Malloc(size)
+            if read_buffer is NULL:
+                raise LibsshSCPException("Memory allocation error")
+
+            # Indicate the transfer is ready to begin
+            libssh.ssh_scp_accept_request(scp)
```
Comment (ganeshrn) [inline src/pylibsshext/scp.pyx:130]: Is return code checking required here as the function can return either `SSH_OK` or `SSH_ERROR `?

File: src/pylibsshext/scp.pyx [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            # Read the file
+            rc = libssh.ssh_scp_read(scp, read_buffer, size)
+            if rc == libssh.SSH_ERROR:
+                PyMem_Free(read_buffer)
```
Comment (ganeshrn) [inline src/pylibsshext/scp.pyx]: This line can be moved to `finally` block.

File: src/pylibsshext/scp.pyx [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                raise LibsshSCPException("Error receiving file data: %s" % self._get_ssh_error_str())
+
+            with open(local_file, "w+") as f:
+                f.write(read_buffer.decode("utf-8"))
+            PyMem_Free(read_buffer)
```
Comment (ganeshrn) [inline src/pylibsshext/scp.pyx]: Same as above, check is read_buffer is allocated before freeing in `finally` block

File: src/pylibsshext/scp.pyx [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                raise LibsshSCPException("Unexpected request: %s" % self._get_ssh_error_str())
+
+        finally:
+            libssh.ssh_scp_close(scp)
+            libssh.ssh_scp_free(scp)
+
+        return libssh.SSH_OK
+
+    def close(self):
+        if self._libssh_session is not NULL:
```
Comment (ganeshrn) [inline src/pylibsshext/scp.pyx]: I think this `close` method should handle freeing of `scp` resources that are still allocated.

**Review by ganeshrn — COMMENTED** [review]
Please add unit tests to cover SCP scenarios.

**Review by ganeshrn — COMMENTED** [review]
Also please update the user_guide https://github.com/ansible/pylibssh/blob/devel/docs/user_guide.rst

File: src/pylibsshext/scp.pyx (hunk 6 above)
Comment (Qalthos) [reply src/pylibsshext/scp.pyx]: Hm, if `_libssh_session` is handled by `session`'s `__dealloc__`, I don't think there is anything left to free as the scp sessions seem to be single-purpose and get freed at the end of a call.

Is there anything I'm missing that needs to be here, or just drop it like `__dealloc__`?

Comment (ganeshrn) [reply src/pylibsshext/scp.pyx]: Yeah if it not required we can drop it

**Comment by webknjaz** [comment]
I'll rebase this to pull in the fixes from devel.

**Review by webknjaz — CHANGES_REQUESTED** [review]
I've added the linter fixes and rebased on top of devel and now it's clear that the test you added (`tests/unit/scp_test.py::test_put`) fails: https://github.com/ansible/pylibssh/pull/157/checks?check_run_id=1504712459#step:13:235. Blocking until solved.

**Comment by Qalthos** [comment]
I updated `user_docs` and split the included files into a few domain-specific versions to keep them a bit easier to follow if someone wants to look at the files directly.

**Review by ganeshrn — APPROVED** [review]
Thanks for working on this.

File: src/pylibsshext/scp.pyx [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+from pylibsshext.errors cimport LibsshSCPException
+from pylibsshext.session cimport get_libssh_session
+
+
+cdef class SCP:
+    def __cinit__(self, session):
+        self.session = session
+        self._libssh_session = get_libssh_session(session)
+
+    def put(self, local_file, remote_file):
```
Comment (webknjaz) [inline src/pylibsshext/scp.pyx:32]: Could you add a docstring?

File: src/pylibsshext/scp.pyx [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+                rc = libssh.ssh_scp_write(scp, PyBytes_AS_STRING(f.read()), file_size)
+                if rc != libssh.SSH_OK:
+                    raise LibsshSCPException("Can't write to remote file: %s" % self._get_ssh_error_str())
+            finally:
+                libssh.ssh_scp_close(scp)
+                libssh.ssh_scp_free(scp)
+
+            return libssh.SSH_OK
+
+    def get(self, remote_file, local_file):
```
Comment (webknjaz) [inline src/pylibsshext/scp.pyx:89]: This needs a docstring too.

File: src/pylibsshext/scp.pyx [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+
+cdef class SCP:
+    def __cinit__(self, session):
+        self.session = session
+        self._libssh_session = get_libssh_session(session)
+
+    def put(self, local_file, remote_file):
+        """
+        Create an SCP channel and send a file to the remote host over that channel
```
Comment (webknjaz) [inline src/pylibsshext/scp.pyx]: PEP 257 fix:
```suggestion
        Create an SCP channel and send a file to the remote host over that channel.
```

File: src/pylibsshext/scp.pyx [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+                    raise LibsshSCPException("Can't write to remote file: %s" % self._get_ssh_error_str())
+            finally:
+                libssh.ssh_scp_close(scp)
+                libssh.ssh_scp_free(scp)
+
+            return libssh.SSH_OK
+
+    def get(self, remote_file, local_file):
+        """
+        Create an SCP channel and retrieve a file from the remote host over that channel
```
Comment (webknjaz) [inline src/pylibsshext/scp.pyx]: PEP257:
```suggestion
        Create an SCP channel and retrieve a file from the remote host over that channel.
```

**Pull request merged**